### PR TITLE
Add OpenEdx E-Commerce app

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

We need to add OpenEdx E-Commerce application to our collection of Arnold's applications.

## Proposal

- [x] integrate first draft release of `edxec`

During this work, I have also noticed that the `lint-git` job was missing from our CI workflow. This is now fixed! :wink:

### CODE REVIEW

File: apps/edxec/templates/services/mysql/svc.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  namespace: "{{ project_name }}"
+spec:
+  ports:
+  - name: "{{ edxec_mysql_port }}-tcp"
+    port: {{ edxec_mysql_port }}
+    protocol: TCP
+    targetPort: {{ edxec_mysql_port }}
+# As commented in the ad hoc endpoint, the endpoint name points to this service
+# so that it does not rely on a deployment configuration when the "env_type" is
+# not trashable. In this case, we use a PostgreSQL cluster outside of
```
Comment (jmaupetit) [inline apps/edxec/templates/services/mysql/svc.yml.j2]:
```suggestion
# not trashable. In this case, we use a MySQL cluster outside of
```

File: apps/edxec/vars/vault/main.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+SECRETS_FILE:
+  DATABASES:
+    default:
+      ENGINE: 'django.db.backends.mysql'
+      NAME: "{{ mysql_credentials.name }}"
+      USER: "{{ mysql_credentials.user }}"
+      PASSWORD: "{{ mysql_credentials.password }}"
+      HOST: "edxec-mysql-{{ deployment_stamp }}"
+      PORT: {{ edxec_mysql_port }}
+      ATOMIC_REQUESTS: true
```
Comment (jmaupetit) [inline apps/edxec/vars/vault/main.yml.j2]: This template should be refactored to fit with current required entries.

File: apps/edxec/templates/services/app/job_00_collectstatic.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            - name: DJANGO_SETTINGS_MODULE
+              value: ecommerce.settings.production
+          command:
+            - "bash"
+            - "-c"
+            - python manage.py collectstatic --noinput && python manage.py compress
+          volumeMounts:
+            - name: edxec-local
+              mountPath: /usr/local/etc/ecommerce
+            - name: edxec-v-media
```
Comment (sampaccoud) [inline apps/edxec/templates/services/app/job_00_collectstatic.yml.j2]: You shouldn't need to mount media to collectstatic?

File: apps/edxec/templates/services/app/job_01_db_migrate.yml.j2 [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+          command:
+            - "bash"
+            - "-c"
+            - python manage.py migrate
+          volumeMounts:
+            - name: edxec-local
+              mountPath: /usr/local/etc/ecommerce
+            - name: edxec-v-media
+              mountPath: /app/media
+            - name: edxec-v-static
```
Comment (sampaccoud) [inline apps/edxec/templates/services/app/job_01_db_migrate.yml.j2]: You shouldn't need to mount statics to run db_migrate?

File: apps/edxec/vars/vault/main.yml.j2 [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# MySQL
+{% set mysql_credentials = databases.mysql | json_query("[?release=='" ~ edxapp_mysql_version ~ "'].databases | [0][?application=='edxec'].{user: user, password: password, name: name} | [0]") %}
```
Comment (sampaccoud) [inline apps/edxec/vars/vault/main.yml.j2]:
```suggestion
{% set mysql_credentials = databases.mysql | json_query("[?release=='" ~ edxec_mysql_version ~ "'].databases | [0][?application=='edxec'].{user: user, password: password, name: name} | [0]") %}
```

File: apps/edxec/templates/services/mysql/secret.yml.j2 [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+apiVersion: v1
+kind: Secret
+metadata:
+  labels:
+    app: edxec
+    service: mysql
+  name: "{{ edxec_mysql_secret_name }}"
+  namespace: "{{ project_name }}"
+data:
+  MYSQL_DATABASE: "{{ EDXEC_VAULT.MYSQL_DATABASE | default('edxapp') | b64encode }}"
```
Comment (sampaccoud) [inline apps/edxec/templates/services/mysql/secret.yml.j2]:
```suggestion
  MYSQL_DATABASE: "{{ EDXEC_VAULT.MYSQL_DATABASE | default('edxec') | b64encode }}"
```

File: apps/edxec/templates/services/mysql/secret.yml.j2 [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  MYSQL_USER: "{{ EDXEC_VAULT.MYSQL_USER | default('edxapp_user') | b64encode }}"
```
Comment (sampaccoud) [inline apps/edxec/templates/services/mysql/secret.yml.j2]:
```suggestion
  MYSQL_USER: "{{ EDXEC_VAULT.MYSQL_USER | default('edxec_user') | b64encode }}"
```

File: apps/edxec/templates/services/app/secret.yml.j2 [hunk 8]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  labels:
+    app: edxec
+    service: app
+  name: "{{ edxec_secret_name }}"
+  namespace: "{{ project_name }}"
+data:
+  local.yml: "{{ EDXEC_VAULT.LOCAL_FILE | default('') | to_nice_yaml | b64encode }}"
```
Comment (lunika) [inline apps/edxec/templates/services/app/secret.yml.j2:10]: Is it allow to have an empty local file ?

File: apps/edxec/templates/services/mysql/secret.yml.j2 (hunk 6 above)
Comment (lunika) [inline apps/edxec/templates/services/mysql/secret.yml.j2]:
```suggestion
  MYSQL_DATABASE: "{{ EDXEC_VAULT.MYSQL_DATABASE | default('edxec') | b64encode }}"
```

File: apps/edxec/templates/services/mysql/secret.yml.j2 (hunk 7 above)
Comment (lunika) [inline apps/edxec/templates/services/mysql/secret.yml.j2]:
```suggestion
  MYSQL_USER: "{{ EDXEC_VAULT.MYSQL_USER | default('edxec_user') | b64encode }}"
```

File: apps/edxec/vars/all/main.yml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+edxec_host: "edxec.{{ project_name}}.{{ domain_name }}"
+
+# -- edxec
+edxec_image_name: "fundocker/edxec"
+edxec_image_tag: "master"
+edxec_django_port: "8000"
+edxec_app_replicas: 1
+edxec_secret_name: "edxec-{{ edxec_vault_checksum | default('undefined_edxec_vault_checksum') }}"
+edxec_activate_http_basic_auth: false
+edxec_yq_image_name: "mikefarah/yq"
```
Comment (lunika) [inline apps/edxec/vars/all/main.yml:13]: do you really think it is needed to set this image name in a variable ?

File: apps/edxec/vars/vault/main.yml.j2 (hunk 5 above)
Comment (lunika) [inline apps/edxec/vars/vault/main.yml.j2]:
```suggestion
{% set mysql_credentials = databases.mysql | json_query("[?release=='" ~ edxec_mysql_version ~ "'].databases | [0][?application=='edxec'].{user: user, password: password, name: name} | [0]") %}
```

File: apps/edxec/templates/services/app/secret.yml.j2 (hunk 8 above)
Comment (jmaupetit) [reply apps/edxec/templates/services/app/secret.yml.j2:10]: Yes.

File: apps/edxec/templates/services/app/job_00_collectstatic.yml.j2 (hunk 3 above)
Comment (jmaupetit) [reply apps/edxec/templates/services/app/job_00_collectstatic.yml.j2]: Good point, thx!

File: apps/edxec/templates/services/app/job_01_db_migrate.yml.j2 (hunk 4 above)
Comment (jmaupetit) [reply apps/edxec/templates/services/app/job_01_db_migrate.yml.j2]: :pray:

File: apps/edxec/vars/all/main.yml (hunk 9 above)
Comment (jmaupetit) [reply apps/edxec/vars/all/main.yml:13]: :man_shrugging: I wanted the tag to be configurable, so by extension, the image is also configurable. Do you think it's useless?

**Comment by jmaupetit** [comment]
@sampaccoud & @lunika I've addressed all your comments. Another review would be warmly appreciated :pray:

File: apps/edxec/vars/all/main.yml (hunk 9 above)
Comment (lunika) [reply apps/edxec/vars/all/main.yml:13]: I think yes but it's not a big deal

File: .circleci/config.yml [hunk 10]
```diff
@@ -362,6 +381,37 @@ jobs:
...
+      - checkout
+      - *attach_workspace
+      - *docker_load
+      - *ci_env
+      - *install_openshift_cluster
+      - *configure_openshift_cluster
+      - *run_openshift_cluster
+
+      - run:
+          name: Test the "edxec" application bootstrapping (with redis enabled)
```
Comment (lunika) [inline .circleci/config.yml]:
```suggestion
          name: Test the "edxec" application bootstrapping
```

File: apps/edxec/templates/services/app/configs/local.yml.j2 [hunk 11]
```diff
@@ -0,0 +1,49 @@
+ALLOWED_HOSTS: {{ (edxec_host | blue_green_hosts).split(',') }}
+DATABASES:
+  default:
+    ENGINE: 'django.db.backends.mysql'
+    # Database name, user and password should be set in the secret
+    HOST: 'edxec-mysql-{{ deployment_stamp }}'
+    PORT: 3306
+    ATOMIC_REQUESTS: true
+# FIXME: is this really required?
```
Comment (madmatah) [inline apps/edxec/templates/services/app/configs/local.yml.j2]: This FIXME is about _ECOMMERCE_URL_ROOT_ ?

It seems to be mandatory, see [here](https://github.com/edx/ecommerce/blob/4b48dc739b42efdd746ede627bffdc7c4ad3d438/e2e/config.py#L20) and [here](https://github.com/edx/ecommerce/blob/4b48dc739b42efdd746ede627bffdc7c4ad3d438/e2e/helpers.py#L48)

And the default value set in the [docker image](https://github.com/openfun/openedx-ecommerce-docker/blob/cfbb22acf4d076fc86c52c5ee53efdecfc448995/docker/files/usr/local/etc/ecommerce/local.yaml#L11) is `http://localhost:8000`

Comment (jmaupetit) [reply apps/edxec/templates/services/app/configs/local.yml.j2]: Well spotted dude! This was a note to self but you did my homework :wink: This is a minor issue for us as the URL may change during the (blue/green) deployment. After checking the source code, it seems that it's only used for end-to-end testing. I'll remove my comment then.

File: .circleci/config.yml [hunk 12]
```diff
@@ -92,6 +92,25 @@ jobs:
           paths:
             - images
 
+  # Check that the git history is clean and complies with our expectations
+  lint-git:
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply .circleci/config.yml:96]: :medal_sports:

**Review by lunika — APPROVED** [review]

**Review by sampaccoud — APPROVED** [review]

**Pull request merged**